source/fuzz/transformation_flatten_conditional_branch.cpp

```diff
@@ -0,0 +1,595 @@
+  // false branch (or the convergence block if there is no false branch).
+  if (last_true_block) {
+    last_true_block->terminator()->SetInOperand(0, {first_false_block_id});
+  }
+
+  // Replace all of the current OpPhi instructions in the convergence block with
+  // OpSelect.
+  ir_context->get_instr_block(convergence_block_id)
+      ->ForEachPhiInst([&condition_operand](opt::Instruction* phi_inst) {
+        phi_inst->SetOpcode(SpvOpSelect);
+        std::vector<opt::Operand> operands;
+        operands.emplace_back(condition_operand);
+        // Only consider the operands referring to the instructions ids, as the
+        // block labels are not necessary anymore.
+        for (uint32_t i = 0; i < phi_inst->NumInOperands(); i += 2) {
+          operands.emplace_back(phi_inst->GetInOperand(i));
+        }
+        phi_inst->SetInOperands(std::move(operands));
+      });
+
+  // Invalidate all analyses
+  ir_context->InvalidateAnalysesExceptFor(opt::IRContext::kAnalysisNone);
+}
+
+protobufs::Transformation TransformationFlattenConditionalBranch::ToMessage()
+    const {
+  protobufs::Transformation result;
+  *result.mutable_flatten_conditional_branch() = message_;
+  return result;
+}
+
+bool TransformationFlattenConditionalBranch::ConditionalCanBeFlattened(
+    opt::IRContext* ir_context, opt::BasicBlock* header,
+    std::set<opt::Instruction*>* instructions_that_need_ids) {
+  uint32_t merge_block_id = header->MergeBlockIdIfAny();
+  assert(merge_block_id &&
+         header->GetMergeInst()->opcode() == SpvOpSelectionMerge &&
+         header->terminator()->opcode() == SpvOpBranchConditional &&
+         "|header| must be the header of a conditional.");
+
+  // Find the first block where flow converges (it is not necessarily the merge
+  // block).
+  uint32_t convergence_block_id = merge_block_id;
+  while (ir_context->cfg()->preds(convergence_block_id).size() == 1) {
+    if (convergence_block_id == header->id()) {
+      // There is a chain of blocks with one predecessor from the header block
+      // to the merge block. This means that the region is not single-entry,
+      // single-exit (because the merge block is only reached by one of the two
+      // branches).
+      return false;
+    }
+    convergence_block_id = ir_context->cfg()->preds(convergence_block_id)[0];
+  }
+
+  auto enclosing_function = header->GetParent();
+  auto dominator_analysis =
+      ir_context->GetDominatorAnalysis(enclosing_function);
+  auto postdominator_analysis =
+      ir_context->GetPostDominatorAnalysis(enclosing_function);
+
+  // Check that this is a single-entry, single-exit region, by checking that the
+  // header dominates the convergence block and that the convergence block
+  // post-dominates the header.
+  if (!dominator_analysis->Dominates(header->id(), convergence_block_id) ||
+      !postdominator_analysis->Dominates(convergence_block_id, header->id())) {
+    return false;
+  }
```

Comment:
It should be OK but make sure to test that the transformation can't be applied to the unreachable selection construct.

Reply:
What do you mean? Should it not be applicable even if a selection construct is unreachable?

Reply:
Maybe we could apply this transformation to an unreachable block but that's a bit tricky. The reason is that domination rules are undefined for them.

If you recall, block A dominates block B if all paths from the entry block to B go through A. In case of unreachable blocks, there is no path from the entry block.

Reply:
This implies that there is no rule that an instruction should dominate all its users.

Reply:
I assume it doesn't get applied now, because nothing dominates anything, right?

Reply:
Yeah. It would still be good to add a test for it, though.

Reply:
I added new tests